test/extensions/filters/http/common/fuzz/filter_fuzz_test.cc

```diff
@@ -0,0 +1,106 @@
+#include <chrono>
+#include <memory>
+
+#include "envoy/config/bootstrap/v2/bootstrap.pb.h"
+#include "envoy/config/filter/http/buffer/v2/buffer.pb.h"
+#include "envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.pb.h"
+
+#include "common/config/utility.h"
+#include "common/protobuf/utility.h"
+
+#include "extensions/filters/http/buffer/buffer_filter.h"
+#include "extensions/filters/http/well_known_names.h"
+
+#include "test/config/utility.h"
+#include "test/extensions/filters/http/common/fuzz/decoder_filter_fuzz.h"
+#include "test/extensions/filters/http/common/fuzz/filter_fuzz.pb.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/mocks/buffer/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/server/mocks.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+
+class UberFilterFuzzer {
+public:
+  UberFilterFuzzer() {
+    ON_CALL(filter_callback_, addStreamDecoderFilter(_))
+        .WillByDefault(
+            Invoke([&](std::shared_ptr<Envoy::Http::StreamDecoderFilter> filter) -> void {
+              filter_ = filter;
+              filter_->setDecoderFilterCallbacks(callbacks_);
+            }));
+  }
+
+  // This executes the methods to be fuzzed.
+  void decode(Http::StreamDecoderFilter* filter, const test::fuzz::HttpData& data) {
+    ENVOY_LOG_MISC(info, "Decoding {} with filter", data.DebugString());
+    bool end_stream = false;
+
+    Http::TestHeaderMapImpl headers = Fuzz::fromHeaders(data.headers());
+    if (data.data().size() == 0 && !data.has_trailers()) {
+      end_stream = true;
+    }
+    filter->decodeHeaders(headers, end_stream);
+
+    for (int i = 0; i < data.data().size(); i++) {
+      if (i == data.data().size() - 1 && !data.has_trailers()) {
+        end_stream = true;
+      }
+      Buffer::OwnedImpl buffer(data.data().Get(i));
+      filter->decodeData(buffer, end_stream);
```

Comment:
I think we need to consider the `decode*` return value, which indicates the stream status. If we're told to stop, we shouldn't provide more data, or we violate the filter's preconditions on these methods.